pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -6587,6 +6664,179 @@ def _preprocess_data(model):
+            for t in model.s_T:
+                above_initial_capacity = False
+                # if the pipeline is defined as bidirectional then the aggregated initial capacity is available in both directions
+                if pipeline[::-1] in model.s_LLA:
+                    initial_capacity = (
+                        model.p_sigma_Pipeline[pipeline].value
+                        + model.p_sigma_Pipeline[pipeline[::-1]].value
+                    )
+                    if (
+                        model.v_F_Piped[pipeline, t].value > initial_capacity
+                        or model.v_F_Piped[pipeline[::-1], t].value > initial_capacity
+                    ):
+                        above_initial_capacity = True
+                else:
+                    initial_capacity = model.p_sigma_Pipeline[pipeline].value
+                    if model.v_F_Piped[pipeline, t].value > initial_capacity:
+                        above_initial_capacity = True
+
+                if above_initial_capacity:
+                    # Add first use to a dictionary
+                    model.infrastructure_firstUse[pipeline] = t
+                    # Get the lead time rounded up to the nearest full time period
+
+                    if model.config.pipeline_cost == PipelineCost.distance_based:
+                        model.infrastructure_leadTime[pipeline] = math.ceil(
+                            model.p_tau_PipelineExpansionLeadTime.value
+                            * model.p_lambda_Pipeline[pipeline].value
+                        )
+                    elif model.config.pipeline_cost == PipelineCost.capacity_based:
+                        # Pipelines lead time may only be defined in one direction
+                        model.infrastructure_leadTime[pipeline] = math.ceil(
+                            max(
+                                model.p_tau_PipelineExpansionLeadTime[i].value,
+                                model.p_tau_PipelineExpansionLeadTime[
+                                    pipeline[::-1], i[2]
+                                ].value,
+                            )
+                        )
+                    break
+
+    # Calculate start build for all infrastructure
+    # Convert the ordered set to a list for easier use
+    s_T_list = list(model.s_T)
+    for key, value in model.infrastructure_firstUse.items():
+        # Find the index of time period in the list
+        finish_index = s_T_list.index(value)
+        start_index = finish_index - model.infrastructure_leadTime[key]
+        if start_index >= 0:
+            model.infrastructure_buildStart[key] = model.s_T.at(start_index + 1)
+        else:
+            model.infrastructure_buildStart[key] = (
+                str(abs(start_index))
+                + " "
+                + model.decision_period.to_string()
+                + "s prior to "
```

Comment:
I recommend adding some logic so that the "s" only gets included when needed. I.e., avoid printing "1 weeks before TXX".

Reply:
Done! 
![image](https://github.com/project-pareto/project-pareto/assets/91741082/006428e8-ec0b-4e87-9c15-0a137960a8d9)